userspace/engine/rule_reader.cpp

```diff
@@ -222,57 +284,81 @@ static void read_item(
 	}
 	else if(item["rule"].IsDefined())
 	{
-		rule_loader::rule_info v;
-		v.ctx = ctx;
-		bool append = false;
+		std::string name;
+
+		// Using tmp context until name is decoded
+		rule_loader::context tmp(item, "rule", "", parent);
+		decode_val(item, "rule", name, tmp);
+
+		rule_loader::context ctx(item, "rule", name, parent);
+		rule_loader::rule_info v(ctx);
+		v.name = name;
+
+		v.append = false;
 		v.enabled = true;
 		v.warn_evttypes = true;
 		v.skip_if_unknown_filter = false;
-		THROW(!decode_val(item["rule"], v.name) || v.name.empty(),
-			"Rule name is empty");
-		if(decode_val(item["append"], append) && append)
+
+		decode_val(item, "append", v.append, ctx, optional);
+
+		if(v.append)
 		{
-			decode_val(item["condition"], v.cond);
-			if (item["exceptions"].IsDefined())
+			decode_val(item, "condition", v.cond, ctx, optional);
+			if(item["condition"].IsDefined())
 			{
-				read_rule_exceptions(item["exceptions"], v);
+				v.cond_ctx = rule_loader::context(item["condition"], "condition", "", ctx);
 			}
+			read_rule_exceptions(item, v, ctx, v.append);
 			loader.append(cfg, v);
 		}
 		else
 		{
-			string priority;
-			bool has_enabled = decode_val(item["enabled"], v.enabled);
-			bool has_defs = decode_val(item["condition"], v.cond)
-					&& decode_val(item["output"], v.output)
-					&& decode_val(item["desc"], v.desc)
-					&& decode_val(item["priority"], priority);
-			if (!has_defs)
+			// If the rule does *not* have any of
+			// condition/output/desc/priority, it *must*
+			// have an enabled property. Use the enabled
+			// property to set the enabled status of an
+			// earlier rule.
+			if (!item["condition"].IsDefined() &&
```

Comment:
I got stuck on this too, but I believe this logic is correct. I believe the goal here is that if a rule does not have any of condition/output/desc/priority, then it *must* have enabled, and the purpose is to simply set enabled for a previously defined rule with the same name.

The old code did this by setting has_defs to true only if all of condition/output/desc/priority were defined, and if !has_defs was true, requiring that enabled was present. I think that's a bit looser than the goal--if a rule defined some but not all of condition/output/desc/priority, has_defs would be false, !has_defs would be true, and enabled would be required. I don't think that's what we want, right?

The new code only checks enabled if *none* of condition/output/desc/priority are present. Otherwise, you jump past the if and try to extract each of condition/output/desc/priority individually (and throw exceptions if any are missing/can't be decoded).

Does this sound right to you too?